### PR TITLE
feat: ESLint class migration

### LINKED ISSUE #10: CLIEngine has been deprecated
Use new ESLint class
https://eslint.org/docs/user-guide/migrating-to-7.0.0#deprecate-cliengine

### PR DESCRIPTION
**Pull request by LKummer**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [x] **metadata update**

### Motivation / Use-Case

[The CLIEngine class has been deprecated.](https://eslint.org/docs/user-guide/migrating-to-7.0.0#deprecate-cliengine)
This PR refactors the plugin to the new ESLint class.

Closes #10.

### Breaking Changes

- Only compatible with ESLint >= 7.
- Incompatible with old CLIEngine options.

For more information on the new ESLint options, [see the ESLint documentation](https://eslint.org/docs/developer-guide/nodejs-api#%E2%97%86-new-eslint-options).

When using deprecated options ESLint will error and suggest how to change the options to the new format.

### Additional Info

Changes introduced in the new ESLint API that affect the plugin:

- ESLint constructor option handling has changed.
  - `options` has changed and is incompatible with `CLIEngine` `options`.
  - `options` is validated to not contain extra properties.
- Linting method is now asynchronous.
- `isPathIgnored()` is now asynchronous.
- `resolveFileGlobPatterns()` has been removed.
- Formatter objects have changed.
  - It is now an object with a `format()` method.
  - Formatter loading is now asynchronous.
  - Formatter loading method is no longer static.
- Results type has changed.

The commits form a list of everything that I have refactored with more detailed information.

I have also updated the peer dependency and the readme.
Please mention if you find anything I have missed.

### CODE REVIEW

File: src/LintDirtyModulesPlugin.js [hunk 1]
```diff
@@ -22,12 +21,8 @@ export default class LintDirtyModulesPlugin {
       return;
     }
 
-    const { cli } = getCLIEngine(this.options);
     const dirtyOptions = { ...this.options };
-    const glob = cli
-      .resolveFileGlobPatterns(dirtyOptions.files)
-      .join('|')
-      .replace(/\\/g, '/');
+    const glob = dirtyOptions.files.join('|').replace(/\\/g, '/');
```
Comment (ricardogobbosouza) [inline src/LintDirtyModulesPlugin.js:25]: We will have problems here. We need to find a solution to `resolveFileGlobPatterns`
https://github.com/webpack-contrib/eslint-webpack-plugin/issues/2
https://github.com/webpack-contrib/eslint-webpack-plugin/commit/d5c8820d9467e8794a4aa3944bf6ded746d79411

Comment (LKummer) [reply src/LintDirtyModulesPlugin.js:25]: Thank you for the links, I will investigate further.

Comment (LKummer) [reply src/LintDirtyModulesPlugin.js:25]: Quick update as I have been silent all week. Thank you for pointing out the issue.

It is indeed causing an issue when watching folder paths with the `lintDirtyModulesOnly` option enabled.

I have added tests and reimplemented the functionality of `resolveFileGlobPatterns`. I am still refactoring and rebasing, it should be ready for review in a day or two.

**Comment by LKummer** [comment]
I have performed the requested change, as well as some more refactoring and a small readme addition.

It is ready for review.

(sorry for the force push, I missed a commit message warning)

**Comment by ricardogobbosouza** [comment]
Thanks @LKummer

**Comment by LKummer** [comment]
Thank you @ricardogobbosouza for merging. Thank you very much @evilebottnawi for the ping.

**Pull request merged**